### PR TITLE
Add `prefer-https` rule

### LINKED ISSUE #971: Rule proposal: `no-http`
Enforce using HTTPS URLs over HTTP.

## Fail

```js
const foo = 'http://sindresorhus.com';

// Hello: http://sindresorhus.com
```

## Pass

```js
const foo = 'https://sindresorhus.com';

// Hello: https://sindresorhus.com
```

---

Alternative name: `prefer-https`. Which one do people prefer?

I'm aware this could be partly solved with the `string-content` rule, but I don't think most people would think of this and it also wouldn't work in comments.

### PR DESCRIPTION
**Pull request by himynameisdave**
Adds `prefer-https` rule.

Checks string literals, as well as comments.

Resolves #971

### CODE REVIEW

**Comment by himynameisdave** [comment]
I set the [`meta.type`](https://eslint.org/docs/developer-guide/working-with-rules#rule-basics) to `'problem'`, not sure if it should be `'suggestion'` instead though.

File: docs/rules/prefer-https.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# Enforce using HTTPS URLs over HTTP
+
+Enforce using HTTPS URLs over HTTP.
```
Comment (sindresorhus) [inline docs/rules/prefer-https.md:3]: This should describe in what places this is enforced.

File: rules/prefer-https.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
+'use strict';
+const getDocumentationUrl = require('./utils/get-documentation-url');
+
+const MESSAGE_ID = 'prefer-https';
+const messages = {
+	[MESSAGE_ID]: 'Prefer `{{replacement}}` over `{{value}}`.'
+};
+
+//	Helper to determine if a string is an unsafe URL.
+const containsHttp = nodeValue => /http:\/\/(www\.)?[\w#%+.:=@~-]{2,256}\.[a-z]{2,4}\b([\w#%&+./:=?@~-]*)/.test(nodeValue);
```
Comment (sindresorhus) [inline rules/prefer-https.js:10]: We don't need to match more after the TLD.

File: rules/prefer-https.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+const create = context => {
+	const sourceCode = context.getSourceCode();
+	const reportPreferHttp = node => {
+		const fixed = node.value.replace('http', 'https');
```
Comment (sindresorhus) [inline rules/prefer-https.js:15]: What if the comment is `// Foo bar http is good: https://asdasd.com`?

I think just replacing `http` substring is too fragile.

File: rules/prefer-https.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+				reportPreferHttp(node);
+			}
+		}
+	};
+};
+
+module.exports = {
+	create,
+	meta: {
+		type: 'problem',
```
Comment (sindresorhus) [inline rules/prefer-https.js:48]: This is not correct.

File: test/prefer-https.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+	valid: [
+		'const foo = \'https://sindresorhus.com\';',
+		'new URL(\'https://sindresorhus.com\');',
+		'const foo = \'http://localhost\''
+	],
+	invalid: [
+		'const foo = \'http://sindresorhus.com\'',
+		'new URL(\'http://sindresorhus.com\')'
+	]
+});
```
Comment (sindresorhus) [inline test/prefer-https.js:14]: You need to add a *lot* more tests.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by fisker** [comment]
Is it possible to disable this rule on comments? Just a question, I never disabled problems in comments.

```js
// eslint-disable-next-line
// http://www.example.com
````

**Comment by himynameisdave** [comment]
Hi @sindresorhus.

I've addressed your comments but am still struggling with the tests. They are currently failing in CI due to some weird module error: `'import' and 'export' may appear only with 'sourceType: module'` (see [here](https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1084/checks?check_run_id=1943974197)).

In general I found it quite difficult to debug this new rule, specifically debugging the output of the `fix`. Not sure what your prescribed method for debugging this would be, but usually for an eslint rule I would just run `eslint ./some-test-file.js --fix` to understand what's happening, although we don't seem to use eslint in that way for `run-rules-on-codebase`.

Another roadblock that I have run into is that, because of the nature of this rule, I needed to disable the rule on the test (because the test cases were being flagged when running `run-rules-on-codebase`). However when I then went to run `npm test`, it complains that there is no definition for the rule!?

All this to say is that I am happy to continue working on this rule, but I am will need some guidance as this is non-obvious to me. It feels much like I am coding in the dark, unable to easily debug what is happening when my rule is run (as well as how the infrastructure that is used to run the tests in this codebase actually works).

**Comment by sindresorhus** [comment]
> They are currently failing in CI due to some weird module error: 'import' and 'export' may appear only with 'sourceType: module' (see here).

You need to rename the file extension to `.mjs`.

**Comment by sindresorhus** [comment]
@fisker can probably give you some tips on debugging `fix`.

**Comment by fisker** [comment]
You can disable it run on codebase by adding it [here](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/c5a43459e881d36e1aa79a7e41a0ec0a05b7e7b7/test/run-rules-on-codebase/lint.mjs#L23), we can enable it when `xo` updated.

**Comment by fisker** [comment]
> It feels much like I am coding in the dark

I'm sorry it makes you feel bad, but `run-rules-on-codebase` is not a required step during implementing, it only need to be done before merge, mentioned in last step here https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/new-rule.md

**Comment by sindresorhus** [comment]
@himynameisdave Still interested in finishing this? :)

**Comment by himynameisdave** [comment]
@sindresorhus yep I'm still interested, although happy to hand this off to someone else if they have more time to work on it. 👍

**Comment by sindresorhus** [comment]
Closing this to let someone else be able to continue it: https://github.com/sindresorhus/eslint-plugin-unicorn/issues/971#issuecomment-933228408

**Pull request closed without merging**